src/libraries/Microsoft.PowerFx.Core/Public/Values/CustomObjectValue.cs

```diff
@@ -11,47 +11,41 @@
 
 namespace Microsoft.PowerFx.Core.Public.Values
 {
+    /// <summary>
+    /// The backing implementation for CustomObjectValue, for example Json, Xml,
+    /// or the Ast or Value system from another language.
+    /// </summary>
     public interface ICustomObject
     {
-        bool IsArray { get; }
-
-        bool IsNull { get; }
-
-        bool IsObject { get; }
-
-        bool IsString { get; }
-
-        bool IsNumber { get; }
-
-        object ToObject();
+        FormulaType Type { get; } // Use ExternalType if the type is incompatible with PowerFx
 
         int GetArrayLength();
 
-        ICustomObject this[int index] { get; }
+        ICustomObject this[int index] { get; } // 0-based index
 
         bool TryGetProperty(string value, out ICustomObject result);
 
         string GetString();
 
         double GetDouble();
+
+        bool GetBoolean();
     }
 
     public class CustomObjectValue : ValidFormulaValue
     {
-        protected readonly ICustomObject _impl;
-
-        public ICustomObject Impl => _impl;
+        public ICustomObject Impl { get; }
 
         internal CustomObjectValue(IRContext irContext, ICustomObject impl)
             : base(irContext)
         {
             Contract.Assert(IRContext.ResultType == FormulaType.CustomObject);
-            _impl = impl;
+            Impl = impl;
         }
 
         public override object ToObject()
         {
-            return _impl.ToObject();
+            return Impl; // Hosts will need to be able to interpret the backing value
```

Comment:
>return Impl; // Hosts will need to be able to interpret the backing value [](http://example.com/codeflow?start=12&length=73)

Let's just throw NotImpl. - that's better than returning something surprising. 

The only right way to impl this is probably as C# dynamic object that's an adapter over ICustomObject. 

We can impl later if somebody actually needs this.

Reply:
Done